Firestore/core/src/firebase/firestore/util/hard_assert.h

```diff
@@ -101,6 +101,43 @@
 namespace firebase {
 namespace firestore {
 namespace util {
+
+typedef void (*FailureHandlerCallback)(const char* file,
+                                       const char* func,
+                                       const int line,
+                                       const std::string& failure_message);
+
+/**
+ * Overrides the default failure handler.
+ *
+ * The default essentially just calls std::terminate. While reasonable for C++,
+ * this isn't optimal for platforms that merely use the C++ core as their
+ * implementation and would otherwise be expected to throw a platform specific
+ * exception.
+ *
+ * @param callback A function that will handle the failure. This function is
+ *     not expected to return. (If it does, std::terminate() will be called
+ *     immediately after it does so.)
+ */
+void SetFailureHandler(FailureHandlerCallback callback);
+
+/**
+ * Default failure handler for non-apple platforms. This should *typically* not
+ * be called directly, though could be called as an implementation detail of a
+ * custom failure handler. eg a failure handler that added logging might look
+ * like:
+ *
+ *   SetFailureHandler([](...) {
+ *     Log("There was an error");
+ *     DefaultFailureHandlerCallback(...);
+ *   });
+ */
+ABSL_ATTRIBUTE_NORETURN void StdioFailureHandlerCallback(
```

Comment:
Comment pruned.

I've left it as a function declaration though. Not necessarily for a good reason: I left it as is mostly to avoid figuring out two separate names for the same thing (one for the function definition, and another for the pointer to it that would be declared here.) In the unlikely event someone *does* use it directly, they'll figure it out pretty quickly. :)  Feel free to push back, in which case, I'll add it in.